8329420: Java 22 (and 23) launcher calls default constructor although main() is static

lahodaj opened the pull request:
Consider code like:
```
public class MainClass {
    public MainClass() {
        System.out.println("Constructor called!");
    }
    public static void main() {
        System.out.println("main called!");
    }
}
```
and compile and run it, with preview enabled, like:
```
$ javac /tmp/MainClass.java 
$ java --enable-preview -classpath /tmp MainClass
Constructor called!
main called!
```

That is wrong, as the `main` method is static, and there is no need to create a new instance of the class.

The reason is that as launcher attempts to invoke the main method, it goes in the following order: 1) static-main-with-args; 2) instance-main-with-args; 3) static-main-without-args; 4) instance-main-without-args. But, for the instance variants, the code first creates a new instance of the given class, and only then attempts to lookup the `main` method, and will pass to the next option when the `main` method lookup fails. So, when invoking static-main-without-args, the current class instance may be created for instance-main-with-args, which will then fail due to the missing `main(String[])` method.

The proposed solution to this problem is to simply first do a lookup for the `main` method (skipping to the next variant when the given main method does not exist, without instantiating the current class).

There is also a relatively closely related problem: what happens when the constructor throws an exception?
```
public class MainClass {
    public MainClass() {
        if (true) throw new RuntimeException();
    }
    public void main() {
        System.out.println("main called!");
    }
}
```

when compiled an run, this produces no output whatsoever:
```
$ javac /tmp/MainClass.java 
$ java --enable-preview -classpath /tmp MainClass
$
```

This is because any exceptions thrown from the constructor are effectively ignored, and the launcher will continue with the next variant. This seems wrong - the exception should be printed for the user, like:
```
$ java --enable-preview -classpath /tmp/ MainClass 
Exception in thread "main" java.lang.RuntimeException
        at MainClass.<init>(MainClass.java:3)
```

This patch proposes to do that by not consuming the exceptions thrown from the constructor, and stop the propagation to the next variant.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329420](https://bugs.openjdk.org/browse/JDK-8329420): Java 22 (and 23) launcher calls default constructor although main() is static (**Bug** - P3)


### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [2022aa5a](https://git.openjdk.org/jdk/pull/18753/files/2022aa5a0d21f930d9b49d1bb0b91ecf7e60ead2)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18753/head:pull/18753` \
`$ git checkout pull/18753`

Update a local copy of the PR: \
`$ git checkout pull/18753` \
`$ git pull https://git.openjdk.org/jdk.git pull/18753/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18753`

View PR using the GUI difftool: \
`$ git pr show -t 18753`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18753.diff">https://git.openjdk.org/jdk/pull/18753.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18753#issuecomment-2051482569)

AlanBateman commented on src/java.base/share/native/libjli/java.c:
```diff
@@ -427,11 +427,11 @@ int
                                  "([Ljava/lang/String;)V");
     CHECK_EXCEPTION_FAIL();
+    jobject mainObject = (*env)->NewObject(env, mainClass, constructor);
+    CHECK_EXCEPTION_NULL_PASS(mainObject);
```
There will be a pending exception if NewObject returns NULL. I think it would be a lot easier to read if this were to just check if mainObject is NULL and return 1. Same comment for the other usage in invokeInstanceMainWithoutArgs. That would avoid CHECK_EXCEPTION_NULL_PASS which is very confusing to see at the use-sites.

lahodaj replied:
Thanks! I've done that here:
https://github.com/openjdk/jdk/pull/18753/commits/2022aa5a0d21f930d9b49d1bb0b91ecf7e60ead2

AlanBateman reviewed (approved):
Thanks for the updated version, much easier to read/maintainer.
Minor nit is that you probably should be a space after each "//" to be consistent with the existing code.

wolfseifert commented:
Does this also fix [JDK-8329581](https://bugs.openjdk.org/browse/JDK-8329581)?

dholmes-ora commented on src/java.base/share/native/libjli/java.c:
```diff
@@ -427,11 +417,16 @@ int
 invokeInstanceMainWithArgs(JNIEnv *env, jclass mainClass, jobjectArray mainArgs) {
     jmethodID constructor = (*env)->GetMethodID(env, mainClass, "<init>", "()V");
     CHECK_EXCEPTION_FAIL();
```
Shouldn't this also not be checked until you know you have an instance method?

dholmes-ora replied:
Yes that is what I meant. If the main method is static then the class need not have any such constructor (unless that is part of the spec for this functionality?).

lahodaj commented:
So, I've pushed an update that:
- attemps to solve [JDK-8329581](https://bugs.openjdk.org/browse/JDK-8329581) by only ignoring `NoSuchMethodError` when looking up methods. Hence any other problem will lead to an error to be printed out, as usually.
https://github.com/openjdk/jdk/pull/18753/commits/dd447e898785a3d7e2def47dd1efcca3db41dd46
- changed the code to first lookup the `main` method, and only then constructors, which is what I think @dholmes-ora was suggesting.
https://github.com/openjdk/jdk/pull/18753/commits/83fabec6e5e8bf6deb9efce22ef6ace408cf3d26
- fixed the comment formatting, as suggested by @AlanBateman
https://github.com/openjdk/jdk/pull/18753/commits/c007f61ebc0660cc6f93e29e71dc635aa50ba538

Please let me know what you think. Thanks!

dholmes-ora reviewed (commented):
Okay on the update I suggested. Can't comment on the overall exception behaviour - not sure what is expected/specified.

jaikiran commented on src/java.base/share/native/libjli/java.c:
```diff
@@ -387,83 +387,97 @@ JLI_Launch(int argc, char ** argv,              /* main argc, argv */
+        // new instance construction failed, don't call the main method,
+        // and don't continue with the next variant;
+        // leave any exception pending, so that it is visible to the caller:
+        return 1;
```
Hello Jan, the comment says "don't continue", yet this returns `1` which represents "continue". Should this return `0` instead?

lahodaj replied:
Uh, you are right of course. Thanks for noticing! Fixed. (I believe the difference here is not currently observable, as this is the last step in the cascade, but may become observable if the order changes.)

wolfseifert commented:
I tested now pull/18753 and pull/18786: both solve both issues JDK-8329420 and JDK-8329581

lahodaj commented:
Superseded by https://github.com/openjdk/jdk/pull/18786.
